lib/puppet/pops/merge_strategy.rb

```diff
@@ -309,8 +309,20 @@ def self.key
     def checked_merge(e1, e2)
       dm_options = { :preserve_unmergeables => false }
       options.each_pair { |k,v| dm_options[k.to_sym] = v unless k == 'strategy' }
-      # e2 (the destination) is dup'ed to avoid that the passed in object mutates
-      DeepMerge.deep_merge!(e1, e2.dup, dm_options)
+      # e2 (the destination) is deep cloned to avoid that the passed in object mutates
+      DeepMerge.deep_merge!(e1, deep_clone(e2), dm_options)
+    end
+
+    def deep_clone(value)
+      if value.is_a?(Hash)
+        result = value.clone
+        value.each{ |k, v| result[k] = deep_clone(v) }
+        result
+      elsif value.is_a?(Array)
+        value.map{ |v| deep_clone(v) }
+      else
+        value
+      end
```

Comment:
Marshah/Unmarshal is more efficient when the structure consists of arbitrary objects and introspection is needed to access the attributes. It's slower when the structure is known to be Data.

The key can something other than a String. It's however never anything that isn't scalar so there's no need to deep copy the key.